File: lintcode/python/1413_tree.py
#!/usr/bin/python -t

# bfs and hash table.
# 两个考点：
# 
# 是否知道怎么根据边来构图和存储图（邻接表）
# 是否知道可以通过 BFS 来确定父子关系

class Solution:
    """
    @param x: The x
    @param y: The y
    @param a: The a
    @param b: The b
    @return: The Answer
    """
    def tree(self, x, y, a, b):
        graph = self.build_graph(x, y)
        parent = self.build_tree(graph)
        
        results = []
        for u, v in zip(a, b):
            if parent[u] == parent[v]:
                results.append(1)
            elif parent[u] == v or parent[v] == u:
                results.append(2)
            else:
                results.append(0)
        return results
    
    def build_graph(self, x, y):
        graph = {}
        for u, v in zip(x, y):
            if u not in graph:
                graph[u] = set()
            if v not in graph:
                graph[v] = set()
            graph[u].add(v)
            graph[v].add(u)
        return graph
        
    def build_tree(self, graph):
        from collections import deque
        visited = set([1])
        queue = deque([1])
        parent = {1: None}
        while queue:
            node = queue.popleft()
            for neighbor in graph[node]:
                if neighbor in visited:
                    continue
                visited.add(neighbor)
                queue.append(neighbor)
                parent[neighbor] = node
        return parent

# hash table
# 一次dfs预处理出所有的信息，然后回答询问即可

class Solution:
    """
    @param x: The x
    @param y: The y
    @param a: The a
    @param b: The b
    @return: The Answer
    """
    def tree(self, x, y, a, b):
        # Write your code here
        n = len(x) + 1
        f = [0 for i in range(n + 1)]
        g = [[] for i in range(n + 1)]
        sons = [{} for i in range(n + 1)]
        for i in range(len(x)):
            print x[i], y[i]
            g[x[i]].append(y[i])
            g[y[i]].append(x[i])
            print g
            
        self.dfs(1, -1, g, f, sons)
        print g, f, sons
        
        ans = [0 for i in range(len(a))]
        for i in range(len(a)):
            if f[a[i]] == f[b[i]]:
                ans[i] = 1
            elif sons[a[i]].has_key(b[i]) or sons[b[i]].has_key(a[i]):
                ans[i] = 2
        return ans
        
    def dfs(self, rt, fa, g, f, sons):
        for x in g[rt]:
            if x == fa:
                continue
            f[x] = rt
            sons[rt][x] = x
            self.dfs(x, rt, g, f, sons)
            

File: lintcode/golang/1413_tree.go

// hash table and bfs

/**
 * @param x: The x
 * @param y: The y
 * @param a: The a
 * @param b: The b
 * @return: The Answer
 */

import "fmt"

func tree (x []int, y []int, a []int, b []int) []int {
    // Write your code here
    graph := make(map[int][]int)
    parent := make(map[int]int)
    m := len(a)
    ret := []int{}

    n := len(x)
    for i , j := 0, 0; i < n && j < n; i, j = i+1, j+1 {
        if xx, ok := graph[x[i]]; ok {
            graph[x[i]] = append(xx, y[j])
        } else {
            graph[x[i]] = []int{y[j]}
        }

        if yy, ok := graph[y[j]]; ok {
            graph[y[j]] = append(yy, x[i])
        } else {
            graph[y[j]] = []int{x[i]}
        }
    }

    fmt.Println("graph: ", graph)

    q := []int{1}
    v := make(map[int]bool)
    v[1] = true
    parent[1] = 0

    for len(q) != 0 {
        level := []int{}

        for _, val := range q {
            for _, nei := range graph[val] {
                if v[nei] == true {
                    continue
                }
                v[nei] = true
                level = append(level, nei)
                parent[nei] = val
            }
        }
        q = level
    }

    fmt.Println("parent: ", parent)

    for i , j := 0, 0; i < m && j < m; i, j = i+1, j+1 {
        if parent[a[i]] == parent[b[j]] {
            ret = append(ret, 1)
        } else if parent[a[i]] == b[j] || parent[b[j]] == a[i] {
            ret = append(ret, 2)
        } else {
            ret = append(ret, 0)
        }
    }

    return ret
}

